test/integration/README.md

```diff
@@ -99,6 +99,9 @@ cluster:
 
 In addition to the existing test framework, which allows for carefully timed interaction and ordering of events between downstream, Envoy, and Upstream, there is now an “autonomous” framework which simplifies the common case where the timing is not essential (or bidirectional streaming is desired). When AutonomousUpstream is used, by setting `autonomous_upstream_ = true` before `initialize()`, upstream will by default create AutonomousHttpConnections for each incoming connection and AutonomousStreams for each incoming stream. By default, the streams will respond to each complete request with “200 OK” and 10 bytes of payload, but this behavior can be altered by setting various request headers, as documented in [`autonomous_upstream.h`](autonomous_upstream.h)
 
+## Common Problems
+- If a response body length does not match the `content-length` header (or there is no header at all), any mock calls to wait for the response completion such as `sendRequestAndWaitForResponse` or `response_->waitForEndStream` could timeout. Also, any asserts that the response was completed such as `EXPECT_TRUE(response_->complete())` would fail. Make sure that the response body length matches the `content-length` header.
```

Comment:
I think this could use a tweak - most requests are "sent" without a content length header, but helper functions add them, for example sendRequestAndWaitForResponse will send an end stream bit, indicating if the response is done (and sometimes causing content length to be added)

I think if you remove "(or there is no header at all)" it'd be more accurate.

Reply:
also timeout -> time out

Reply:
I faced this problem when I was sending a request using `makeHeaderOnlyRequest` and encoding the response using `upstream_request_->encodeXXX()`. Even when I set `end_stream`, for e.g. by using `upstream_request_->encodeData(_, true)`, the call for `response->waitForEndStream` times out. This happened because I did not include a `content-length` to the encoded headers at all. That's why I added the "no header at all" part.

I guess whether the `content-length` header is added or not depends on the protocol, so it should not be something we depend on. I know this because when I was missing the `content-length` header, the integration test passed for some protocols and timed out for others.

I'd say its better for whoever is writing the test to add a correct `content-length` header, even if some functions/protocols would implicitly add it anyway. That's why I also chose the wording "it could time out" rather than "it will time out".